File: leetcode/1047.remove-all-adjacent-duplicates-in-string/go/removeduplicates.go
package removeduplicates

type simpleStack struct {
	s []rune
}

func (s *simpleStack) Pop() bool {
	if len(s.s) == 0 {
		return false
	}

	s.s = s.s[:len(s.s)-1]
	return true
}

func (s *simpleStack) Peek() (rune, bool) {
	if len(s.s) == 0 {
		return '0', false
	}

	return s.s[len(s.s)-1], true
}

func (s *simpleStack) isEmpty() bool {
	return len(s.s) == 0
}

func (s *simpleStack) Push(v rune) {
	s.s = append(s.s, v)
}

func (s *simpleStack) Export() string {
	return string(s.s)
}

func removeDuplicates(s string) string {
	stack := simpleStack{}

	for _, v := range s {
		prev, ok := stack.Peek()

		if ok && prev == v {
			stack.Pop()
		} else {
			stack.Push(v)
		}
	}

	return stack.Export()
}

File: leetcode/1047.remove-all-adjacent-duplicates-in-string/ts/removeduplicates.ts
const popStack = (stack: string[]): void => {
    stack.pop()
}

const peekStack = (stack: string[]): string | null => {
    if (stack.length === 0) return null

    return stack[stack.length - 1]
}

const pushStack = (stack: string[], v: string): void => {
    stack.push(v)
}

const isEmptyStack = (stack: string[]): boolean => {
    return stack.length === 0
}

const removeDuplicates = (s: string): string => {
    const stack: string[] = []

    for (const ch of s) {
        if (!isEmptyStack(stack) && peekStack(stack) === ch) popStack(stack)
        else pushStack(stack, ch)
    }

    return stack.join("")
}

export default removeDuplicates
